src/daft-local-execution/src/sinks/grouped_aggregate.rs

```diff
@@ -0,0 +1,403 @@
+            let state = state.get_or_insert_with(|| SinglePartitionAggregateState {
+                partially_aggregated: vec![],
+                unaggregated: vec![],
+                unaggregated_size: 0,
+            });
+            state.partially_aggregated.push(p);
+        }
+        Ok(())
+    }
+
+    fn execute_partition_then_agg(
+        inner_states: &mut [Option<SinglePartitionAggregateState>],
+        input: Arc<MicroPartition>,
+        params: &GroupedAggregateParams,
+    ) -> DaftResult<()> {
+        let partitioned =
+            input.partition_by_hash(params.group_by.as_slice(), inner_states.len())?;
+        for (p, state) in partitioned.into_iter().zip(inner_states.iter_mut()) {
+            let state = state.get_or_insert_with(|| SinglePartitionAggregateState {
+                partially_aggregated: vec![],
+                unaggregated: vec![],
+                unaggregated_size: 0,
+            });
+            if state.unaggregated_size + p.len() >= GroupedAggregateState::PARTIAL_AGG_THRESHOLD {
+                let unaggregated = std::mem::take(&mut state.unaggregated);
+                let aggregated =
+                    MicroPartition::concat(unaggregated.iter().chain(std::iter::once(&p)))?.agg(
+                        params.partial_agg_exprs.as_slice(),
+                        params.group_by.as_slice(),
+                    )?;
+                state.partially_aggregated.push(aggregated);
+                state.unaggregated_size = 0;
+            } else {
+                state.unaggregated_size += p.len();
+                state.unaggregated.push(p);
+            }
+        }
+        Ok(())
+    }
+
+    fn execute_partition_only(
+        inner_states: &mut [Option<SinglePartitionAggregateState>],
+        input: Arc<MicroPartition>,
+        params: &GroupedAggregateParams,
+    ) -> DaftResult<()> {
+        let partitioned =
+            input.partition_by_hash(params.group_by.as_slice(), inner_states.len())?;
+        for (p, state) in partitioned.into_iter().zip(inner_states.iter_mut()) {
+            let state = state.get_or_insert_with(|| SinglePartitionAggregateState {
+                partially_aggregated: vec![],
+                unaggregated: vec![],
+                unaggregated_size: 0,
+            });
+            state.unaggregated_size += p.len();
+            state.unaggregated.push(p);
+        }
+        Ok(())
+    }
+}
+
+struct SinglePartitionAggregateState {
+    partially_aggregated: Vec<MicroPartition>,
+    unaggregated: Vec<MicroPartition>,
+    unaggregated_size: usize,
+}
+
+enum GroupedAggregateState {
+    Accumulating {
+        inner_states: Vec<Option<SinglePartitionAggregateState>>,
+        strategy: Option<AggStrategy>,
+    },
+    Done,
+}
+
+impl GroupedAggregateState {
+    // This is the threshold for when we should aggregate any accumulated unaggregated partitions
+    const PARTIAL_AGG_THRESHOLD: usize = 10_000;
```

Comment:
let's have both of these as part of the execution config?  it would be nice to be able to tune these in runtime rather than having to recompile